### PR TITLE
pac builder actions section

### PR DESCRIPTION
**Pull request by ibnc**
part of #6857

Waiting on the config repo id field to be added

### CODE REVIEW

File: server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import * as Buttons from "views/components/buttons";
+import css from "./styles.scss";
+
+interface Attrs {
+  material: Stream<Material>;
+  pluginId: Stream<string>;
+  loc?: LocationHandler;
+}
+
+export interface LocationHandler {
```
Comment (marques-work) [inline server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: As this is duplicated from `views/pages/pipelines/actions.tsx`, might be good to extract `LocationHandler` and `WindowLocation` into `helpers/`. Mostly an abstraction used in testing.

File: server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    event.stopPropagation();
+    this.location.go("/go/pipelines");
+  }
+
+  onSave(pluginId: string, material: Material, event: Event): void {
+    event.stopPropagation();
+    event.preventDefault();
+    this.clearErrors();
+
+    const configRepo = new ConfigRepo(material.name(), pluginId, material);
```
Comment (marques-work) [inline server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: This is likely going to be provided by a separate user input field soon. There's a PR that will remove the "material name" and "destination" fields from the PaC material form, just FYI.

Comment (ibnc) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: @marques-work link to that PR so I can have a better understanding of what's gonna change?

Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: This is the one I'm talking about:

https://github.com/gocd/gocd/pull/6847

Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: @ibnc I also pushed a conflict fix to your branch

Comment (ibnc) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: @marques-work Okay, I see what you mean. Yeah, as I mentioned in the description of this PR, this PR is waiting on the pac id input field. I figured the material name would work fine, but it's probably better to have a separate field. Regardless, this PR won't be mergeable until that's been finished.

Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: @ibnc You could just add a field and label it with "Name this repository" or something, maybe?

Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: @ibnc If you were referring to the live validating fields, they already exist. The incantation you want to use is probably `<IdentifierInputField/>` located in `server/webapp/WEB-INF/rails/webpack/views/components/forms/common_validating_inputs.tsx`

Comment (ibnc) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: @marques-work That's what I'm planning on doing today. Albeit I didn't know about that field

File: server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  onSave(configRepo: ConfigRepo, event: Event): void {
+    event.stopPropagation();
+    event.preventDefault();
+    this.clearErrors();
+
+    if (configRepo.isValid()) {
+      ConfigReposCRUD.create(configRepo).then((result: ApiResult<any>) => {
+        result.do((s) => {
+          this.location.go("/go/admin/config_repos");
```
Comment (marques-work) [inline server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: we can now link directly to the repo:

```typescript
this.location.go(`/go/admin/config_repos#!${encodeURIComponent(configRepo.id())}`);
```

Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pac/actions.tsx]: Might want to make that a paramter-optional `SparkRoute` entry?

`function configRepoPage(repoId?: string): string {}`

File: server/webapp/WEB-INF/rails/webpack/views/pages/pipelines_as_code.tsx [hunk 4]
```diff
@@ -106,17 +111,23 @@ export class PipelinesAsCodeCreatePage extends Page {
...
                 this.material(cloneMaterialForPaC(vm.material));
+                this.configRepo().material(this.material());
               }
             }}
           />
 
           <MaterialEditor material={this.material()} hideTestConnection={syncConfigRepoWithMaterial()} scmOnly={true} showLocalWorkingCopyOptions={false} disabled={syncConfigRepoWithMaterial()}/>
         </UserInputPane>
         <div class={css.verifyDefsInMaterial}>
+          <IdentifierInputField label="Identifier" property={this.configRepo().id} errorText={this.configRepo().errors().errorsForDisplay("id")} />
```
Comment (marques-work) [inline server/webapp/WEB-INF/rails/webpack/views/pages/pipelines_as_code.tsx]: I think we've found that users aren't clear about "Identifier" as it sounds like something we should generate, not them. I know this is basically a primary key used to reference a configrepo, but for other similar concepts, we call this a "name" (e.g., pipeline name). Could we change this label to read "Name this repository"? Or similar?

Comment (ibnc) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pipelines_as_code.tsx]: @marques-work the only issue is the validation error will say `id must be present` 

We can change the error message in the constructor, but that'll effect other forms as well.

Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/pipelines_as_code.tsx]: @ibnc Hmm. Does this error come from the config repo model? specifically, `validatePresenceOf()`. Let's make the change to the label, and I'll look to see where we might need to fix the validation message elsewhere

**Comment by ibnc** [comment]
> Looks good now. Maybe squash some of the minor commits (or all if you want) and remove the WIP commit message?
> 
> After the 19.8.0 release, we can merge.

Pretty much done. I'm having trouble removing "WIP" from the commit message because of the merge. It won't rebase smoothly, and I don't want to screw up the history on accident...

**Comment by ibnc** [comment]
@marques-work even with only doing the reword, it wasn't happy about the merge commit lol

**Pull request merged**